### PR TITLE
[SPARK-46395][DOCS] Assign Spark configs to groups for use in documentation

### PR DESCRIPTION
**Pull request by nchammas**
### What changes were proposed in this pull request?

Summary:
- Eliminate the need to manually maintain HTML tables for Spark configurations.
- Using the approach introduced here, we also:
  - Ensure that internal configs are not accidentally documented publicly.
  - Ensure that configs are documented publicly just as they are documented in the code.
- Reorganize the sections in the SQL tuning page slightly so that related tuning techniques are clearer.
- Make it possible for users to hyperlink directly to config entries in our documentation.

This change enables us to assign the various SQL configs to groups that can then be referenced automatically in our documentation. This will replace the large number of manually maintained HTML tables with references to generated tables.

For example, the SQL performance tuning page has a section on data caching that references two configs. They are in an HTML table that has to be [manually maintained][1].

[1]: https://github.com/apache/spark/blob/7db85642600b1e3b39ca11e41d4e3e0bf1c8962b/docs/sql-performance-tuning.md?plain=1#L37-L56

With this new system, we group the configs in `SQLConf.scala` according to how we'd like to display them in our documentation. For example, we can assign some configs to the `sql-tuning-caching-data` group:

```scala
  val COMPRESS_CACHED = buildConf("spark.sql.inMemoryColumnarStorage.compressed")
    ...
    .withTag("sql-tuning-caching-data")
```

Then we can reference this `sql-tuning-caching-data` config group from within `sql-performance-tuning.md` as follows:

```liquid
{% include_api_gen _generated/config_tables/sql-tuning-caching-data.html %}
```

This pulls in an automatically generated HTML table that includes all the configs tagged `sql-tuning-caching-data`.

In addition to eliminating the need to manually maintain HTML tables of configs, this PR also fixes some problems with the current documentation:
1. It ensures that internal configs are not accidentally published in the user-facing documentation, because [internal configs do not get exported][exported] to documentation. For example, `spark.sql.files.openCostInBytes` is an [internal config][internal], but it is currently documented publicly in the SQL tuning guide.
2. It ensures that the public documentation of a config matches exactly what is written in the code. For example, the public documentation for `spark.sql.autoBroadcastJoinThreshold` is [missing][missing] the following line that is [in the code][in-the-code]:
    > and file-based data source tables where the statistics are computed directly on the files of data.

[exported]: https://github.com/apache/spark/blob/84e27dcb8b4bebc6544934e0e299e228f6579e7f/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L5650
[internal]: https://github.com/apache/spark/blob/ccc436d829cd0b07088e2864cb1ecc55ab97a491/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L1810-L1811
[missing]: https://spark.apache.org/docs/3.5.0/sql-performance-tuning.html#other-configuration-options
[in-the-code]: https://github.com/apache/spark/blob/14a933bbe2eb1c71988f475036735f07b2e1fa6a/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L596-L597

This PR also adds anchors to each config in the generated HTML tables, so that people can link directly to specific configurations in the documentation.

This PR builds on the work done in #27459, and is related to the work done in #28274.

### Why are the changes needed?

This makes Spark configuration documentation much easier to maintain and more useful to users.

### Does this PR introduce _any_ user-facing change?

Yes, it alters some of the user-facing documentation.

### How was this patch tested?

Manually built the documentation and viewed it in my browser.

Here's a screenshot of the generated docs: [sql-performance-tuning.html 📸][screenshot]

I also reviewed the generated documentation for `configuration.html` but didn't screenshot it because it's too long.

### Was this patch authored or co-authored using generative AI tooling?

No.

[screenshot]: https://github.com/apache/spark/assets/1039369/f05a64e0-fb4b-4d88-99b2-22d03a19b9a2

### CODE REVIEW

File: docs/sql-performance-tuning.md [hunk 1]
```diff
@@ -34,30 +34,9 @@ memory usage and GC pressure. You can call `spark.catalog.uncacheTable("tableNam
...
-  <td>10000</td>
-  <td>
-    Controls the size of batches for columnar caching. Larger batch sizes can improve memory utilization
-    and compression, but risk OOMs when caching data.
-  </td>
-  <td>1.1.1</td>
-</tr>
-
-</table>
+{% include_relative generated-sql-config-table-caching-data.html %}
```
Comment (nchammas) [inline docs/sql-performance-tuning.md:37]: This diff demonstrates the main benefit of this PR.

**Review by nchammas — COMMENTED** [review]
@HyukjinKwon @cloud-fan @gatorsmile - What do you think? Tagging you because you last reviewed the documentation machinery I am modifying here.

File: docs/README.md [hunk 2]
```diff
@@ -46,8 +46,6 @@ $ cd docs
 $ bundle install
 ```
 
-Note: If you are on a system with both Ruby 1.9 and Ruby 2.0 you may need to replace gem with gem2.0.
```
Comment (nchammas) [inline docs/README.md:49]: Both Ruby 1 and 2 are EOL.

File: sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala [hunk 3]
```diff
@@ -72,6 +72,14 @@ private[sql] object PythonSQLUtils extends Logging {
...
 
+  def listAllSQLConfigsWithTags():
+    Array[(String, String, String, String, Array[String])] = {
+    val conf = new SQLConf()
+    conf.getAllDefinedConfsWithTags.map(c =>
+      Tuple5(c._1, c._2, c._3, c._4, c._5.toArray)
+    ).toArray
+  }
+
   private def listAllSQLConfigs(): Seq[(String, String, String, String)] = {
```
Comment (nchammas) [inline sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala:83]: We can remove some of these other methods if we go with this PR's approach of assigning arbitrary tags to configs.

That includes:
- `listAllSQLConfigs`
- `listRuntimeSQLConfigs`
- `listStaticSQLConfigs`

File: core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala [hunk 4]
```diff
@@ -197,6 +199,13 @@ private[spark] case class ConfigBuilder(key: String) {
     this
   }
 
+  def withTag(tag: String): ConfigBuilder = {
```
[the remark this answers is not in the document]
Comment (nchammas) [reply core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala:202]: We could also do away with custom group names and instead select certain prefixes for use as documentation groups, like `spark.sql.cbo`, `spark.sql.statistics`, etc.

However, this won't work for groupings that don't align with config name prefixes, like the breakdown of runtime vs. static configurations that was added in #28274.

**Comment by nchammas** [comment]
There are about 55 HTML tables of Spark configs (both SQL and non-SQL) across our documentation. They probably span a few thousand lines of HTML. I believe they can all be replaced with automatically generated HTML tables.

I intend to do that if this PR gets buy-in from committers.

**Comment by HyukjinKwon** [comment]
(will try to take a look around next week)

**Comment by nchammas** [comment]
Thank you, Hyukjin. (I hope when I edited [this comment][1] it didn't re-ping you. That wasn't my intention.)

[1]: https://github.com/apache/spark/pull/44300#pullrequestreview-1776188809

**Comment by nchammas** [comment]
Btw @HyukjinKwon if there is anything I can do to make this PR easier to review, I'm all ears. The key change is the introduction of `withTag` in `ConfigBuilder.scala`, and the rest of the PR flows from that.

**Comment by nchammas** [comment]
@HyukjinKwon - How are you feeling about this PR? Shall I recruit another committer to take a look instead?

I have the time and motivation to migrate all of our config documentation to use this new system. As I noted in a previous comment, there are around 55 config tables across our documentation that span several thousand lines of manually maintained HTML. They can all be eliminated with this approach. I think it would be a big win for the maintainability of our documentation.

**Comment by nchammas** [comment]
Silence on a PR usually means there is something wrong with it, so I am proactively trying to figure out how to move this idea forward.

The possibilities I am considering are:
1. Committers disapprove of the abstract idea of automating the generation of config tables.
2. Committers approve of the abstract idea but disapprove of the implementation proposed in this PR.
3. Committers approve of the abstract idea and may approve of the implementation proposed here, but find it difficult to review.
4. Committers are constrained (due to time, energy, other priorities, etc.) and cannot participate at this time.

In the case of 2, I have proposed a completely different approach in #44756 that does not touch Spark core.

In the case of 3, I pushed #44755, which is a stripped down version of this PR which should be easier to review.

In the case of 1, I need some feedback on why this is a bad idea so I can drop this effort. As you can see, I believe it's a good idea and am eager to push it forward.

In the case of 4, please forgive me for repeatedly pinging this PR and being annoying. 😅

**Comment by bjornjorgensen** [comment]
@nchammas there seams to be a lot (some) of PR's that have been approved now, but have not been merged to master. So I think it have something to do with Hollidays and so on..

This is not the first PR for fixing the docs. Can an email to @dev be a good thing her?

**Pull request closed without merging**